Reduce OpenMP ifdefs; add receiver calculation onto GPU streams

davschneller opened the pull request:
* rework analytical boundary condition a bit; avoid re-allocating far too often
* encapsule OpenMP calls more idiomatically
* add the on-fault and off-fault receiver calculation onto some GPU streams

vikaskurapati commented on src/DynamicRupture/Output/Builders/PickPointBuilder.h:
```diff
@@ -13,6 +13,7 @@
 #include "ReceiverBasedOutputBuilder.h"
 
 #include <Common/Iterator.h>
+#include <Parallel/Runtime/Stream.h>
```
Not related to this PR, but as we discussed in another PR, if we can slowly ensure that all new headers added are of the consistent style of `"xxx.h"`, it would be nice. I understand if you don't want to do it here, that is why this is left as a comment, and not as a request changes.

vikaskurapati commented on src/DynamicRupture/Output/OutputManager.cpp:
```diff
@@ -30,6 +30,7 @@
 #include "Memory/Tree/Lut.h"
 #include "ResultWriter/FaultWriterExecutor.h"
 #include "SeisSol.h"
+#include <Parallel/Runtime/Stream.h>
```
Same as the header comment above. I won't write more of these comments to avoid redundancy.

vikaskurapati commented on src/Solver/TimeStepping/TimeCluster.cpp:
```diff
@@ -81,7 +79,7 @@ TimeCluster::TimeCluster(unsigned int clusterId,
           maxTimeStepSize, timeStepRate, seissolInstance.executionPlace(clusterData->size())),
       // cluster ids
-      usePlasticity(usePlasticity), seissolInstance(seissolInstance),
+      usePlasticity(usePlasticity), seissolInstance(seissolInstance), streamRuntime(4),
```
Just for my clarification, what does the 4 here refer to? Why exactly 4?

vikaskurapati reviewed (commented):
Looks good to me for OpenMP ifdefs, and receiver calculations parts. But the PR has code which is more than what is written in the PR description. Please add the tests which you used to verify the initial conditions' implementations, and easi boundary conditions so that it is complete.

davschneller commented on docs/initial-condition.rst:
```diff
@@ -6,19 +6,68 @@
+However, it requires easi 1.5.0 or higher to function correctly.
+
+As a caveat, the Easi initial condition currently does not support the analytical boundary condition.
+
```
Ok, I'll do that.

davschneller replied:
Actually, it wasn't added in this PR but already last year. With the plan of someday (possibly in a 2.0.0) replacing all the current hard-coded initial conditions. Cf. https://github.com/SeisSol/precomputed-seissol/pull/19 .

It'll still need to be extended to the analytical BCs; I'm gonna do that probably next. (after #1405 and #1401 are done most likely)

davschneller commented on src/Kernels/LinearCK/Local.cpp:
```diff
@@ -78,31 +79,32 @@ void Local::setGlobalData(const CompoundGlobalData& global) {
+    for (std::size_t s = 0; s < multisim::NumSimulations; ++s) {
+      auto slicedBoundaryDofs = multisim::simtensor(boundaryDofs, s);
+      initConditions->at(s % initConditions->size())
+          ->evaluate(time, nodesVec, NodeCount, localData.material(), slicedBoundaryDofs);
```
Compiled it mainly—the focus here was to update the implementation and avoid passing a vector down the line. I'm not sure if anyone has even tested just "default" initial conditions with multiple simulations recently (w/o imposing them at the boundary). Which should work, however.

(maybe some test to add by you?)

I'm gonna refactor the analytical BC again anyway next (to add the easi boundaries for them as well); so I'm going to wait with adding a proper test case until then from my side—then I can also add a fused case for all of it if it hasn't happened until then already.

davschneller replied:
Yes, that sounds good. I'll add it (yes, noone besides you (or me in exceptions) will work on it—but just in case we're far too busy to ever address it and leave it to our sucessors; though I don't hope that we'll get to that).

davschneller commented:
Tiny nitpickity: they _were_ referenced in the first bullet point. :)

If I were to split up all the PRs even more cleanly, we might have ended up with 50 open PRs at once.
I would have liked to do so; but since the whole reviewing process was a bit slower than I had originally anticipated, it would've become really difficult to manage all open PRs (it is already a bit tedious now).

vikaskurapati commented:
> Tiny nitpickity: they _were_ referenced in the first bullet point. :)
> 

Sorry, seemed to have missed it.

> If I were to split up all the PRs even more cleanly, we might have ended up with 50 open PRs at once. I would have liked to do so; but since the whole reviewing takes time, it would've become really difficult to manage all open PRs (it is already a bit tedious now).

It seems that you misunderstood what I am trying to say. I don't want you to split up all the PRs. What I requested is to add whatever you used to test these new features somewhere so that anyone who modifies something related to those can test later to ensure they don't break anything. 

That being said, I am approving this now. You can consider the comments if you wish to modify something; otherwise, Thanks!

davschneller commented:
Ok—sorry for misunderstanding; and thanks for the approval. I've addressed the comments—and plan to extend the easi boundary conditions pretty soon.

Also, I get what you mean—and I truly hope that we can accomplish a mostly-verified SeisSol version in a reasonable timeframe soon-ish. (even though we're all rather busy)
